### PR TITLE
Stops machine frames and circuitboards referring to manipulators as micro-manipulators

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

1. Gives stock parts a base_name variable and sets it to manipulator for micro-manipulators
2. Reorders sprites of stock parts in stock_parts.dmi to be together and in order of tier. [Image](https://i.imgur.com/RuIU2oP.png)
3. Makes component part lists on machine frames and circuitboards nicer and makes them identical between the two
4. Improves code for generating these lists
5. Adds numtotext() proc which these lists use

## Why It's Good For The Game

1. Solves the issue where examining machines says they require a micro-manipulator when any manipulator will do
2. Makes sprites easier to find
3. Wows players by using actual words for numbers in generated parts lists
4. Cleans up ancient code
5. This proc can be used for other things, e.g. station names (which I have now incorporated it into)

## Changelog

:cl:
fix: Machine frames and circuit boards no longer call manipulators micro-manipulators when examined
code: Improves code in constructable_frame.dm and circuitboard.dm
/:cl:

### CODE REVIEW

**Comment by SuperNovaa41** [comment]
Soul removal

File: code/modules/research/techweb/all_nodes.dm [hunk 1]
```diff
@@ -14,6 +14,7 @@
 		"basic_matter_bin",
 		"basic_micro_laser",
 		"basic_scanning",
+		"basic_manipulator",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:17]: ABC order

Comment (cacogen) [reply code/modules/research/techweb/all_nodes.dm:17]: Oh yeah I was gonna correct that but then I forgot

Comment (Mothblocks) [reply code/modules/research/techweb/all_nodes.dm]: It's mostly my fault for not setting up linters on this yet 😅

**Comment by Farquaar** [comment]
Any plans to change micro-laser as well?

**Comment by cacogen** [comment]
Well no because every tier is a micro-laser and what changes is the power. Although maybe I could fix manipulators by similarly making every tier a micro-manipulator instead.

image

**Comment by newfren** [comment]
the phrase "femto micro" is very silly to me. you could just make machines request "manipulators" instead

**Comment by cacogen** [comment]
My argument that it doesn't matter even if the prefix is supposed to refer to size is:

1) They aren't microscopic (or nanoscopic)
2) They're all the same size
2) There's no such thing as picoscopic or femtoscopic
3) They all have the same description that only refers to them as "tiny little"

**Comment by cacogen** [comment]
> A micromanipulator is a device which is used to physically interact with a sample under a microscope.

So they're all micro manipulators, objects capable of manipulating microscopic things. The higher tiers of micro manipulator are just capable of manipulating even smaller things. 

You could shorten them to things like femto-manipulator the way we do at the moment, but that means the base object is considered a micro-manipulator and the other tiers aren't. 

That causes the problem when examining machine frames where it asks for a micro-manipulator when any tier of manipulator will do. I first solved the problem by changing the base micro-manipulator to just manipulator, but I don't like that solution either.

I'll probably add a base name string to stock parts that machine frames can pull from instead if it exists.

File: code/modules/research/designs/stock_parts_designs.dm [hunk 2]
```diff
@@ -113,10 +113,10 @@
 	departmental_flags = DEPARTMENTAL_FLAG_ENGINEERING | DEPARTMENTAL_FLAG_SCIENCE
 
 //Maipulators
-/datum/design/micro_mani
-	name = "Micro Manipulator"
+/datum/design/basic_manipulator
+	name = "Basic Manipulator"
 	desc = "A stock part used in the construction of various devices."
-	id = "micro_mani"
+	id = "basic_manipulator"
```
Comment (Fikou) [inline code/modules/research/designs/stock_parts_designs.dm:119]: you sure you still want this change in if you took a different solution?

Comment (cacogen) [reply code/modules/research/designs/stock_parts_designs.dm:119]: Yeah, it puts them next to the other tier 1 parts in the stock part list and names them similarly, which helps new players understand they're tier 1.

**Comment by cacogen** [comment]
pls merge

**Comment by AlinaStarkova** [comment]
"Wows players by using actual words for numbers in generated parts lists"

not sure about this one, numbers as words are harder to notice, and for those who arent fluent in english theres no reason to force the use of words to see the numbers, most games that denote costs usually have the word and number together, like "one (1) manipulator"

admittedly it really doesnt matter since nobody actually reads that, as they just fill a BRPD and just spam machines with it

**Comment by cacogen** [comment]
I meant that in kind of an ironic way. I still think it looks good though.

The game uses words as numbers elsewhere so this feels more like concern trolling because you didn't like something I wrote more than anything.

I can't speak for other players but I use these all the time, particularly if I don't want to waste mats printing a bunch of parts. But the scenario you presented doesn't even make sense because you'd need to check the machine for things like glass sheets or cable coils. Would you prefer they were just removed and people had to consult the wiki again?

**Comment by AlinaStarkova** [comment]
> I meant that in kind of an ironic way. I still think it looks good though.
> 
> The game uses words as numbers elsewhere so this feels more like concern trolling because you didn't like something I wrote more than anything.
> 
> I can't speak for other players but I use these all the time, particularly if I don't want to waste mats printing a bunch of parts. But the scenario you presented doesn't even make sense because you'd need to check the machine for things like glass sheets or cable coils. Would you prefer they were just removed and people had to consult the wiki again? 

your being a bit of an ass just for me bringing up a legitimate concern. not everyone who plays is native English speakers and doing "one (1)" is a pretty standard accessibility thing most games do, i dont think thats concern trolling to mention. and yes actually, I do think the other areas of the game that only use wordage for numbers hurt accessibility. no, I wouldn't prefer it be removed entirely, i never even suggested anywhere close to such, no need for sarcasm. 

as for glass sheets or cable coils, sure, thats checked if your confused why you cant finish the machine, but lets be real, you bped the parts in, screwdriver, if it doesnt work you slap glass and cable at it and try again. if it still doesnt work then you check whats missing. thats basically how most engineers build machines.

File: code/__HELPERS/text.dm [hunk 3]
```diff
@@ -809,6 +809,31 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
 		else
 			return "[number]\th"
 
+/proc/numtotext(number)
```
Comment (Ryll-Ryll) [inline code/__HELPERS/text.dm:812]: Should specify either in the name (inttotext?) or the autodoc that this only supports integers

File: code/game/machinery/constructable_frame.dm [hunk 4]
```diff
@@ -31,39 +31,38 @@
...
-		var/requires = "It requires"
-
-		for(var/i = 1 to req_components.len)
-			var/tname = req_components[i]
-			var/amt = req_components[tname]
-			if(amt == 0)
+		var/list/nice_list = list()
+		for(var/atom/component as anything in req_components)
+			if(!ispath(component))
+				continue
```
Comment (Ryll-Ryll) [inline code/game/machinery/constructable_frame.dm:37]: Should probably throw a stack_trace() here so we know what's wrong

**Pull request merged**